### PR TITLE
Add type annotation in `_cuda_types.py`

### PR DESCRIPTION
**Pull request by asi1024**
```
$ mypy cupy cupyx cupy_backends --enable-error-code attr-defined | grep cupyx/jit
```

### CODE REVIEW

File: cupyx/jit/_cuda_types.py [hunk 1]
```diff
@@ -1,84 +1,100 @@
...
         self.child_type = child_type
         self.ndim = ndim
 
 
 class CArray(ArrayBase):
-    def __init__(self, dtype, ndim, is_c_contiguous, index_32_bits):
+
+    def __init__(
+            self,
+            dtype: numpy.dtype,
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupyx/jit/_cuda_types.py]: Fix to allow npt.DTypeLike type argument for consistency.

[the remark this answers is not in the document]
Comment (kmaehashi) [reply cupyx/jit/_cuda_types.py]: DTypeLike is a new feature in v1.20 but we still need to support v1.19~ for v11 and v1.18~ for v10.

[the remark this answers is not in the document]
Comment (kmaehashi) [reply cupyx/jit/_cuda_types.py]: ah, but we are going to drop v1.19 in v11, so I think it should be fine. I'll send a PR to drop NumPy 1.19.

[the remark this answers is not in the document]
Comment (kmaehashi) [reply cupyx/jit/_cuda_types.py]: #6735

File: cupyx/jit/_cuda_types.py [hunk 2]
```diff
@@ -1,84 +1,100 @@
...
 
-    def __str__(self):
+    def __str__(self) -> str:
         ctype = get_typename(self.dtype)
         c_contiguous = get_cuda_code_from_constant(self._c_contiguous, bool_)
         index_32_bits = get_cuda_code_from_constant(self._index_32_bits, bool_)
         return f'CArray<{ctype}, {self.ndim}, {c_contiguous}, {index_32_bits}>'
 
-    def __eq__(self, other):
+    def __eq__(self, other: TypeBase) -> bool:  # type: ignore[override]
```
Comment (kmaehashi) [inline cupyx/jit/_cuda_types.py:97]: I think `other` should be `object`.
https://stackoverflow.com/questions/37557411/why-does-defining-the-argument-types-for-eq-throw-a-mypy-type-error

File: cupyx/jit/_cuda_types.py [hunk 3]
```diff
@@ -118,31 +139,31 @@ def declvar(self, x, init):
...
+    def __init__(self, types: Sequence[TypeBase]) -> None:
         self.types = types
 
-    def __str__(self):
+    def __str__(self) -> str:
         types = ', '.join([str(t) for t in self.types])
         return f'thrust::tuple<{types}>'
 
-    def __eq__(self, other):
+    def __eq__(self, other: TypeBase) -> bool:  # type: ignore[override]
```
Comment (kmaehashi) [inline cupyx/jit/_cuda_types.py:158]: ditto

**Review by kmaehashi — APPROVED** [review]

**Pull request merged**